### ISSUE TITLE
Make TPOT Robust for Noisy Datasets

### ISSUE THREAD

**Issue by CSNoyes** [rhiever/tpot#514]
I recently raised #513, which had to do with the random_state in TPOT being a (false) contributing factor to the selection of the top pipeline. For the problem referenced there, the problem was not intractable (logistic regression with some preprocessing, a deterministic model) gave 59% classification accuracy, which is actually great.

For highly noisy datasets, I would recommend adding an option to TPOT to randomize the random_state used for the pipelines at each step in the pipeline optimizations (perhaps at each generation or before the scoring function is called). This would make TPOT much more robust for high noise problems, as it would help to prevent good learners from being marginally overtaken by overfit (through the selection of the random_state) learners. For example, a RF with a 'good' random_state beat logistic regression for my problem in #513, but logistic regression was actually learning and generalizing better, and I wouldn't have known that without filtering all of the top genomes in the final population manually by testing them with different random weights.

**Comment by weixuanfu2016** [comment]
@rhiever Do you like this idea? I think a piratical way for adding random seed is to add `random_state` into built-in configuration of operators, like `'random_state': [42, 84, 100,...]` for operators with this parameters. Or we can add this function into `self._add_terminals()`

**Comment by rhiever** [comment]
The reason we evaluate all TPOT pipelines with a specific `random_state` is so the pipeline will give a consistent score every time they are evaluated. If pipelines gave different scores every time they were evaluated, then it would be problematic for us to rank pipelines because, as you said, some pipelines can just get "lucky" with a particular seed. Also, having pipelines with a fixed seed is advantageous because we can use a cache whenever we encounter the same pipeline in the future, which means we can save time by avoiding re-evaluations of pipelines.

I think the general consensus from the ML community is that if an algorithm/pipeline's performance is highly dependent on its random seed, then it is not a very good nor generalizable algorithm/pipeline. As such, we felt comfortable fixing every TPOT pipeline's seed and count on k-fold CV to properly assess the generalization performance of every pipeline.

I'm actually quite surprised that the pipeline with a RF you mentioned in #513 works dramatically differently with different `random_state` values for the RF. I didn't think RFs were affected that much by their `random_state`, especially with >=100 trees. Can you verify that it wasn't the cross-validation procedure that was causing inconsistencies in the pipeline performance?

**Comment by rhiever** [comment]
Ping. Any progress on this, @CSNoyes?

**Comment by CSNoyes** [comment]
Hey, sorry for the radio silence. The RF was adversely affected, but not to a degree that would *usually* be significant (1.5% accuracy). However, there are many times when a lot of KNN's are tried in the initial population in which some random combination or parameters leads to a good accuracy. However, that result is truly due only to randomness, and fails completely with a different random_state.

The KNN example is a more insidious one, since if you have a small and noisy training set (let's say ~1000 samples), the mechanism of evolution in TPOT will carry that randomly good KNN through and continue to mutate it as it's still the top genome. Changing the random_state at each iteration and re-calculating the cross_val_score fixes this issue. The randomly good genomes fall away very quickly and it made my problem tractable. It also requires only a few lines changed in base.py. I would recommend adding it as an option for small, noisy datasets, or I can maintain a separate fork with these capabilities.

**Comment by rhiever** [comment]
It sounds like this issue could be driven more by the data split moreso than the random state of the algorithms. KNNs, for example, aren't stochastic algorithms and don't use the random state.

Are you shuffling your dataset beforehand? Are you performing stratified splits before passing the data to TPOT?

**Comment by rhiever** [comment]
Ping @CSNoyes

**Comment by saddy001** [comment]
> I think the general consensus from the ML community is that if an algorithm/pipeline's performance is highly dependent on its random seed, then it is not a very good nor generalizable algorithm/pipeline.

Do you have any links for further reading? I think it is a very interesting question if an estimator, that has proven good during CV with a certain random state is not good at generalization. Because I thought we use CV to achieve exactly this.

**Comment by rhiever** [comment]
@saddy001: We do indeed use CV to get an estimate of generalization performance. AFAIK, I don't know of any paper that explicitly explores the issue of random seeds and generalizability. However, it makes inherent sense that an algorithm's performance shouldn't heavily depend on its random seed. Random seeds are designed for reproducibility and aren't intended to be another "knob" (i.e., parameter) to tune in the algorithm.

**Comment by saddy001** [comment]
Currently I'm seeing the same behaviour as CSNoyes, especially for time series problems. 
As we can only give one sample split to TPOT at the beginning of the optimization process, the pipeline seems to overfit this special split during fitting, even if there are many splits. 
In the end, this results in a good score reported by TPOT, but when evaluating the best pipeline on other splits, the performance is much worse. I think a simple reproducing example could be built, if you wanted, I could give it a try.

There may be two options to overcome this: 
1. Changing the random_state at each iteration as suggested by CSNoyes
2. Randomizing splits at each iteration. 

To me, option 1 seems more easy. As an additional option to TPOT, it wouldn't hurt the overall performance of TPOT.

**Comment by saddy001** [comment]
PS: Option 2. is [what sklearn does](http://scikit-learn.org/stable/modules/cross_validation.html#computing-cross-validated-metrics) when given a cross validation split generator, that is evaluated at each iteration.

**Comment by JoshHeitzman** [comment]
I like the idea of changing the random_state each generation.  Independent of that I've been thinking about being able to specify the population on a per generation basis as I'm thinking it may be useful to be able to keep shrinking it or even dynamically expand/contract it based on the number of pipelines in the pareto front.  Putting the two together leads to the idea of a per generation configuration either provided statically upfront or that is retrieved from user supplied function before each generation is executed.

**Comment by weixuanfu** [comment]
TPOT currently stores average cv scores of evaluated pipelines with a fixed random_state and skip evaluating duplicated pipelines to save computational time. Changing the random_state or cv splits each generation need re-evaluate those duplicated pipelines.

I like the idea about a per generation configuration.

**Comment by JoshHeitzman** [comment]
Yes, changing the random_state would definitely incur the cost of increased compute in exchange for the expected benefit of increased generalization robustness of the final model.  Definitely something that should be should be opted into rather than a default.
